### PR TITLE
support no-loc for <xref:uid> and @uid

### PR DESCRIPTION
**Pull request by AngryBerryMS**
[AB#532858](https://ceapex.visualstudio.com/Engineering/_workitems/edit/532858)

* external xref are treated as not localizable in this PR
* internal xref is localizable when the uid is from a md file with a localizable title as the display text
* internal xref if not localizable when the uid is from a SDP yml file in this PR (will be covered in another pr)
* xref-not-found uids will be covered in another pr
* \<xref href=.../\> will be covered in another pr

### CODE REVIEW

File: docs/specs/escape.yml [hunk 1]
```diff
@@ -162,11 +162,11 @@ outputs:
           <p>
             Link to <a href='a.json'>text</a>
             Link to <a href='a.json'>&lt;some text &amp;&gt;\n</a>
```
Comment (herohua) [inline docs/specs/escape.yml:164]: These two seem should have same localization behavior:
```
    Link to <xref href="a&b?displayProperty=description" />
    Link to <xref:a&b?displayProperty=description>
```

Do you plan to fix this in a later dev task?

File: docs/specs/escape.yml [hunk 2]
```diff
@@ -162,11 +162,11 @@ outputs:
...
-            Link to <a href='a.json'>&lt;some text &amp;&gt;\n</a>
-            Link to <a href='a.json'>&lt;some text &amp;&gt;\n</a>
+            Link to <a class=\"no-loc\" href='a.json'>&lt;some text &amp;&gt;\n</a>
+            Link to <a class=\"no-loc\" href='a.json'>&lt;some text &amp;&gt;\n</a>
             Link to @a&amp;c
```
Comment (herohua) [inline docs/specs/escape.yml:167]: This should be `<a class="no-loc">@a&amp;c</a>`

File: docs/specs/escape.yml [hunk 3]
```diff
@@ -162,11 +162,11 @@ outputs:
...
             Link to &lt;xref:a%26&gt;
```
Comment (herohua) [inline docs/specs/escape.yml:168]: Also here should be `<a class="no-loc">&lt;xref:a%26&gt;</a>`

**Comment by herohua** [comment]
Can you add a comment on what scenarios would be handled in later PRs?

File: src/docfx/build/xref/InternalXrefMapBuilder.cs [hunk 4]
```diff
@@ -93,7 +93,9 @@ private void Load(ErrorBuilder errors, ListBuilder<InternalXrefSpec> xrefs, File
         var monikers = _monikerProvider.GetFileLevelMonikers(errors, file);
         var xref = new InternalXrefSpec(metadata.Uid, _documentProvider.GetSiteUrl(file), file, monikers);
 
-        xref.XrefProperties["name"] = new Lazy<JToken>(() => new JValue(string.IsNullOrEmpty(metadata.Title) ? metadata.Uid : metadata.Title.Value));
+        JToken token = new JValue(string.IsNullOrEmpty(metadata.Title) ? metadata.Uid : metadata.Title.Value);
+        var localizable = !string.IsNullOrEmpty(metadata.Title);
+        xref.XrefProperties["name"] = new Lazy<(JToken token, bool localizable)>((token, localizable));
```
Comment (herohua) [inline src/docfx/build/xref/InternalXrefMapBuilder.cs:98]: can we move into Lazy?

File: src/docfx/build/xref/InternalXrefSpec.cs [hunk 5]
```diff
@@ -15,19 +15,27 @@ internal record InternalXrefSpec(SourceInfo<string> Uid, string Href, FilePath D
...
+          && propertyValue.Value is string internalStr
+          ? (internalStr, property.Value.localizable)
+          : default;
     }
 
-    public string? GetName() => GetXrefPropertyValueAsString("name");
+    public string? GetName() => GetXrefPropertyValueAsString("name").propertyValue;
+
+    public (string? name, bool localizable) GetXrefPropertyOfName()
+    {
```
Comment (herohua) [inline src/docfx/build/xref/InternalXrefSpec.cs:36]: how about only keep one of these two?

File: src/docfx/build/xref/XrefResolver.cs [hunk 6]
```diff
@@ -51,7 +51,7 @@ internal class XrefResolver
         _internalXrefMap = new(BuildInternalXrefMap);
     }
 
-    public (Error? error, string? href, string display, FilePath? declaringFile) ResolveXrefByHref(
+    public (Error? error, string? href, string display, FilePath? declaringFile, bool localization) ResolveXrefByHref(
```
Comment (herohua) [inline src/docfx/build/xref/XrefResolver.cs:54]: >localization [](http://example.com/codeflow?start=86&length=12)

localizable

File: src/docfx/build/xref/ExternalXrefSpec.cs [hunk 7]
```diff
@@ -30,16 +30,16 @@ internal record ExternalXrefSpec : IXrefSpec
 
     FilePath? IXrefSpec.DeclaringFile => null;
 
-    public string? GetXrefPropertyValueAsString(string propertyName)
+    public (string? propertyValue, bool localizable) GetXrefPropertyValueAsString(string propertyName)
```
Comment (herohua) [inline src/docfx/build/xref/ExternalXrefSpec.cs:33]: >string? propertyValue, bool localizable [](http://example.com/codeflow?start=12&length=39)

consider invent `Localizable<T>` to encapsulate `(string?, bool)` and `(JToken?, bool)`

File: src/docfx/build/xref/XrefResolver.cs [hunk 8]
```diff
@@ -79,15 +79,35 @@ internal class XrefResolver
         var (xrefError, xrefSpec, resolvedHref) = ResolveXrefSpec(new SourceInfo<string>(uid, href.Source), referencingFile, inclusionRoot);
         if (xrefError != null || xrefSpec is null || string.IsNullOrEmpty(resolvedHref))
         {
-            return (xrefError, null, alt ?? "", null);
+            return (xrefError, null, alt ?? "", null, !string.IsNullOrEmpty(alt));
         }
 
-        var displayPropertyValue = displayProperty is null ? null : xrefSpec.GetXrefPropertyValueAsString(displayProperty);
+        var (displayPropertyValue, localizableOfDisplayProperty) =
```
Comment (yufeih) [inline src/docfx/build/xref/XrefResolver.cs:85]: We should probably limit the scope to the `title` property in conceptual pages only. All the other xref syntax is mostly used for references that aren't localizable. Most use of `displayProperty` is to reference non-localizable strings,  e.g. [`displayProperty=nameWithType`](https://github.com/dotnet/docs-archive-pr/blob/8b02460ce2f2775e490ac629959d0a29c51a5f04/docs/framework/windows-identity-foundation/namespace-mapping-between-wif-3-5-and-wif-4-5.md#wif-35-to-wif-45-namespace-map)

File: src/docfx/build/xref/InternalXrefSpec.cs (hunk 5 above)
Comment (AngryBerryMS) [reply src/docfx/build/xref/InternalXrefSpec.cs:35]: Done

File: src/docfx/build/xref/XrefResolver.cs (hunk 6 above)
Comment (AngryBerryMS) [reply src/docfx/build/xref/XrefResolver.cs]: Fixed.

File: src/docfx/build/xref/ExternalXrefSpec.cs (hunk 7 above)
Comment (AngryBerryMS) [reply src/docfx/build/xref/ExternalXrefSpec.cs]: Done

**Comment by AngryBerryMS** [comment]
> 

Done

File: src/docfx/build/xref/InternalXrefMapBuilder.cs (hunk 4 above)
Comment (AngryBerryMS) [reply src/docfx/build/xref/InternalXrefMapBuilder.cs]: Done

File: src/docfx/build/xref/XrefResolver.cs (hunk 8 above)
Comment (AngryBerryMS) [reply src/docfx/build/xref/XrefResolver.cs]: Yes. We limit the scope here 
image

File: src/docfx/build/xref/InternalXrefSpec.cs [hunk 9]
```diff
@@ -15,19 +15,27 @@ internal record InternalXrefSpec(SourceInfo<string> Uid, string Href, FilePath D
 
     public string? SchemaType { get; init; }
 
-    public Dictionary<string, Lazy<JToken>> XrefProperties { get; } = new Dictionary<string, Lazy<JToken>>();
+    public Dictionary<string, Lazy<(JToken token, bool localizable)>> XrefProperties { get; } =
```
[the remark this answers is not in the document]
Comment (AngryBerryMS) [reply src/docfx/build/xref/InternalXrefSpec.cs]: Done

File: src/docfx/lib/HtmlUtility.cs [hunk 10]
```diff
@@ -225,7 +225,7 @@ public static void TransformLink(ref HtmlToken token, MarkdownObject? block, Fun
         ref HtmlReader reader,
         ref HtmlToken token,
         MarkdownObject? block,
-        Func<SourceInfo<string>?, SourceInfo<string>?, bool, (string? href, string display)> resolveXref)
+        Func<SourceInfo<string>?, SourceInfo<string>?, bool, (string? href, string display, bool localizable)> resolveXref)
```
[the remark this answers is not in the document]
Comment (AngryBerryMS) [reply src/docfx/lib/HtmlUtility.cs]: Done

File: src/docfx/lib/XrefResolvingResult.cs [hunk 11]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Docs.Build;
+
+internal readonly struct XrefResolvingResult
```
Comment (yufeih) [inline src/docfx/lib/XrefResolvingResult.cs:6]: This structure is actually representing a link:
```suggestion
internal readonly struct XrefLink
```

Comment (yufeih) [reply src/docfx/lib/XrefResolvingResult.cs:6]: Also make `XrefResolver.ResolveXrefByHref` method return `(Error, XrefLink)` to make this type propagate fullstack.

File: src/docfx/build/xref/ExternalXrefSpec.cs [hunk 12]
```diff
@@ -30,16 +30,16 @@ internal record ExternalXrefSpec : IXrefSpec
 
     FilePath? IXrefSpec.DeclaringFile => null;
 
-    public string? GetXrefPropertyValueAsString(string propertyName)
+    public LocInfo<string?> GetXrefPropertyValue(string propertyName)
```
[the remark this answers is not in the document]
Comment (AngryBerryMS) [reply src/docfx/build/xref/ExternalXrefSpec.cs]: Fixed

File: src/docfx/lib/XrefResolvingResult.cs (hunk 11 above)
Comment (AngryBerryMS) [reply src/docfx/lib/XrefResolvingResult.cs]: Done

File: src/docfx/build/xref/InternalXrefSpec.cs [hunk 13]
```diff
@@ -15,7 +15,9 @@ internal record InternalXrefSpec(SourceInfo<string> Uid, string Href, FilePath D
 
     public string? SchemaType { get; init; }
 
-    public Dictionary<string, Lazy<JToken>> XrefProperties { get; } = new Dictionary<string, Lazy<JToken>>();
+    public bool IsNameLocalizable { get; set; }
+
+    public Dictionary<string, Lazy<JToken>> XrefProperties { get; } = new();
```
Comment (live1206) [inline src/docfx/build/xref/InternalXrefSpec.cs:20]: Should `IsLocalizable` be combined with the specific Xref property?

File: src/docfx/build/xref/XrefResolver.cs [hunk 14]
```diff
@@ -113,10 +120,12 @@ internal class XrefResolver
         var (error, xrefSpec, href) = ResolveXrefSpec(uid, referencingFile, inclusionRoot, monikers);
         if (error != null || xrefSpec == null || href == null)
         {
-            return (error, null, "", null);
+            return (error, new XrefLink(null, "", null, false));
         }
+
+        var localizable = xrefSpec is InternalXrefSpec internalXrefSpec && internalXrefSpec.IsNameLocalizable;
         _fileLinkMapBuilder.AddFileLink(inclusionRoot, referencingFile, xrefSpec.Href, uid.Source);
```
Comment (herohua) [inline src/docfx/build/xref/XrefResolver.cs:127]: consolidate this with line 93

**Review by herohua — APPROVED** [review]
:shipit:

File: src/docfx/build/xref/InternalXrefSpec.cs (hunk 13 above)
Comment (AngryBerryMS) [reply src/docfx/build/xref/InternalXrefSpec.cs:20]: We can consider title only up to now.

File: src/docfx/build/xref/XrefResolver.cs (hunk 14 above)
Comment (AngryBerryMS) [reply src/docfx/build/xref/XrefResolver.cs:126]: Done

File: src/docfx/lib/XrefLink.cs [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+{
+    public readonly string? Href;
+
+    public readonly string Display;
+
+    public readonly FilePath? DeclaringFile;
+
+    public readonly bool Localizable;
+
+    public XrefLink()
```
[the remark this answers is not in the document]
Comment (AngryBerryMS) [reply src/docfx/lib/XrefLink.cs]: removed

**Review by live1206 — APPROVED** [review]

**Pull request merged**